File: api/question-types-data.py
#!/usr/bin/env python3

import os
import sys
import json
import tempfile
import urllib.parse
from http.server import BaseHTTPRequestHandler

# Add the parent directory to sys.path to import our modules
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import our existing analysis functions
try:
    from generate_wordcloud import (
        load_and_process_csv, 
        generate_question_types_bar_chart,
        ensure_nltk_data
    )
except ImportError as e:
    print(f"Import error: {e}")

# Import cloud storage functions
def get_data_from_blob():
    """Get data from Vercel Blob storage"""
    try:
        # Try to import and use cloud storage
        sys.path.append('/var/task/api')  # Vercel function path
        from cloud_storage import getDataWithInit
        result = getDataWithInit()
        if result.get('success') and result.get('data'):
            return result['data'], 'vercel-blob'
    except Exception as e:
        print(f"Failed to get data from Vercel Blob: {e}")
    
    # Fallback to local file for development
    try:
        csv_path = os.environ.get('CSV_FILE_PATH', 'data/demo_feedback.csv')
        if os.path.exists(csv_path):
            import pandas as pd
            df = pd.read_csv(csv_path)
            return df.to_dict('records'), 'local-file'
    except Exception as e:
        print(f"Failed to load local file: {e}")
    
    return None, None

def convert_data_to_csv(data):
    """Convert data array to CSV string"""
    if not data:
        return ""
    
    import pandas as pd
    df = pd.DataFrame(data)
    return df.to_csv(index=False)

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            # Ensure NLTK data is available
            ensure_nltk_data()
            
            print("Question types analysis API called")
            
            # Get data from Vercel Blob or local file
            data, data_source = get_data_from_blob()
            
            if not data:
                self.send_response(500)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                error_response = {
                    'success': False,
                    'error': 'No data available',
                    'details': 'Could not load data from Vercel Blob or local files'
                }
                self.wfile.write(json.dumps(error_response).encode())
                return
            
            print(f"Loaded {len(data)} records from {data_source}")
            
            # Create temporary CSV file
            with tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False) as temp_file:
                csv_content = convert_data_to_csv(data)
                temp_file.write(csv_content)
                temp_csv_path = temp_file.name
            
            try:
                # Generate question types analysis
                print(f"Processing CSV file for question types analysis: {temp_csv_path}")
                
                # Ensure public directory exists
                os.makedirs('public', exist_ok=True)
                
                # Generate question types bar chart (this function handles the analysis internally)
                chart_data = generate_question_types_bar_chart(temp_csv_path)
                
                if not chart_data:
                    raise Exception("Question types analysis failed - no data returned")
                
                print(f"Question types analysis completed successfully")
                
                # Return success response with chart data
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.send_header('Access-Control-Allow-Origin', '*')
                self.end_headers()
                
                response = {
                    'success': True,
                    'message': 'Question types analysis completed successfully',
                    'data': chart_data,
                    'recordCount': len(data),
                    'dataSource': data_source,
                    'analysis': 'question-types'
                }
                
                self.wfile.write(json.dumps(response).encode())
                
            finally:
                # Clean up temporary file
                try:
                    os.unlink(temp_csv_path)
                except:
                    pass
                    
        except Exception as e:
            print(f"Error in question types analysis: {str(e)}")
            
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            
            error_response = {
                'success': False,
                'error': 'Failed to generate question types analysis',
                'details': str(e),
                'dataSource': data_source if 'data_source' in locals() else 'unknown'
            }
            
            self.wfile.write(json.dumps(error_response).encode())
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers() 

File: api/question-types-data.js
const { spawn } = require('child_process');
const fs = require('fs');

export default async function handler(req, res) {
  // Set CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'GET') {
    return res.status(405).json({ success: false, error: 'Method not allowed' });
  }

  try {
    console.log('Question types API called');
    
    const csvPath = process.env.CSV_FILE_PATH || 'data/demo_feedback.csv';
    
    // Set up Python environment for question types
    const env = {
      ...process.env,
      CSV_FILE_PATH: csvPath,
      ANALYSIS_TYPE: 'question_types'
    };

    return new Promise((resolve, reject) => {
      const pythonProcess = spawn('python3', ['generate_wordcloud.py', '--question-types=true'], {
        env,
        cwd: process.cwd()
      });

      let output = '';
      let errorOutput = '';

      pythonProcess.stdout.on('data', (data) => {
        output += data.toString();
        console.log('Python stdout:', data.toString());
      });

      pythonProcess.stderr.on('data', (data) => {
        errorOutput += data.toString();
        console.error('Python stderr:', data.toString());
      });

      pythonProcess.on('close', (code) => {
        console.log('Python process exited with code:', code);
        
        if (code === 0) {
          const expectedFile = 'public/wordcloud_question_types.png';
          const imagePath = '/wordcloud_question_types.png';
          
          if (fs.existsSync(expectedFile)) {
            const result = {
              success: true,
              imagePath: imagePath,
              timestamp: new Date().toISOString(),
              mode: 'question_types'
            };
            console.log('Returning result:', result);
            res.status(200).json(result);
            resolve();
          } else {
            console.error('Expected file not found:', expectedFile);
            res.status(500).json({
              success: false,
              error: 'Question types analysis was not generated',
              expectedFile,
              pythonOutput: output,
              pythonError: errorOutput
            });
            resolve();
          }
        } else {
          console.error('Python script failed with code:', code);
          res.status(500).json({
            success: false,
            error: 'Failed to generate question types analysis',
            code,
            output,
            errorOutput
          });
          resolve();
        }
      });

      pythonProcess.on('error', (error) => {
        console.error('Failed to start Python process:', error);
        res.status(500).json({
          success: false,
          error: 'Failed to start question types analysis process',
          details: error.message
        });
        resolve();
      });
    });

  } catch (error) {
    console.error('API error:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Internal server error'
    });
  }
} 